### ISSUE TITLE
Interpretation scopes should be declared

### ISSUE THREAD

**Issue by jashug** [coq/coq#7120]
Currently, a scope is automatically created with the first of `Notation "a" := b : scope`, `Delimit scope with key`, or `Bind Scope scope with T`.
(are there other ways to declare a scope?)
In contrast, `Open Scope scope` does not create scope.

@RalfJung commented (https://github.com/coq/coq/pull/6247#issuecomment-377202990) that the current implicit binding is awkward, since typos will silently create a new scope. I agree.
Sometimes you also want to declare a sequence of notations in a new scope:
```
Notation "+" := ( 1 ) : AA.
Open Scope AA.
Notation "*" := ( + ) : AA.
```
Here, `Open Scope AA` has to come *second* in the list of commands: it makes more sense to have it at the top, but `AA` hasn't been declared yet, so you get an error.

I propose adding a new command `Declare Scope scope` which is equivalent to the commands `Delimit Scope scope with placeholder. Undelimit scope.`,
and adding a warning whenever a new scope is created other than by `Declare Scope`, with the intention to turn the warning into an error after a suitable deprecation period.

**Comment by herbelin** [comment]
Ah, I actually also started to do it. Are you already done on your side?

**Comment by Zimmi48** [comment]
@jashug I'll invite you to the Coq organization so that you can self-assign issues to signal when you're working on them.

**Comment by jashug** [comment]
@herbelin No, I just started, and I can't work on this immediately. Please go ahead.

If there was discussion about the design, here would be a good place to do that.

Figuring out a reasonable transition strategy is what stymied me. The straightforward approach gives a warning when files are required as well, which is inconvenient for libraries worried about compatibility: assuming this is added in 8.9, they can't add `Declare Scope` until 8.10, but all their users (which may switch to 8.9 before 8.10 is released) get a bunch of warnings.

If we let `Declare Scope` be repeated, end users could declare scopes that are then populated by the libraries they require. If `Declare Scope` can't be repeated (which I think is natural and more useful for catching bugs), then when the library adds `Declare Scope`, users will break (though it's a simple fix; maybe this isn't such a problem).

I suppose unsetting the warning before `Require` and setting it after would be another way for users to deal with it...

I'm not sure how much to worry about any of the above; and we can always refine when warnings are emitted later.

**Comment by herbelin** [comment]
> Figuring out a reasonable transition strategy is what stymied me. The straightforward approach gives a warning when files are required as well, which is inconvenient for libraries worried about compatibility: assuming this is added in 8.9, they can't add `Declare Scope` until 8.10, but all their users (which may switch to 8.9 before 8.10 is released) get a bunch of warnings.

> I suppose unsetting the warning before `Require` and setting it after would be another way for users to deal with it...

Indeed. Conclusion is that we should deactivate the warning in `Require`, right?

> If we let `Declare Scope` be repeated, end users could declare scopes that are then populated by the libraries they require. If `Declare Scope` can't be repeated (which I think is natural and more useful for catching bugs), then when the library adds `Declare Scope`, users will break (though it's a simple fix; maybe this isn't such a problem).

> Notation clashes hit a similar limitation of the warning infrastructure w.r.t. `Require`. So we may want to improve that.

Indeed. I made a test in the standard library and it happens a couple of time that files refer to the same scope without one being directly dependent on the other (e.g. `Programs.Utils` and `Programs.Basics`, or `Classes.Morphisms` and `Classes.CMorphisms`). My opinion on these given examples is that it is legitimate to have two `Declare Scope`which are eventually joined. But maybe they are examples where it would not be good.

On the other side, still in the standard library, there are independent files which defined a scope of same name without this scope intended to be the same. This is the case of `pair_scope` which is defined in all of `ssrfun.v`, `FMapAVL`, `MSetAVL`, `ZNatPairs.v`. Altogether, these files define 2 different notations for `fst` and `snd` and notations for two distinct notions of pairs (in `prod` and in a specific AVL structure). But how to ensure that two distinct libraries will never define the same scope? Should scopes be qualifiable by the library name to protect from unexpected collisions? 

> If there was discussion about the design, here would be a good place to do that.

Probably, we shall need a `Local Declare Scope` (and asking that the successive `Notation` and co have a `Local` flag too).

Another issue, which shall have a priori to be treated in #496, is that ML printers for numerals declare notations in a scope. We have to decide whether the scope is expected to be defined in advance prior to the corresponding `Declare ML Module` or if the `Declare ML Module` does the `Declare Scope` itself (with some subtleties about undoing the `Declare ML Module` needing to undo also the `Declare Scope`).

A preliminary work is [here](https://github.com/herbelin/github-coq/tree/master%2Bexplicit-declare-scope-command) but none of the four potential points above are addressed (deactivating warning at `Require`, qualified scope names, scopes in primitive numerals, local declaration of a scope).

**Comment by jashug** [comment]
@herbelin 

This turned into a bit of a wall.
TL;DR: After spending another couple hours pondering, I think
* we should deactivate the warning in Require
* qualified scope names is probably too ambitious but is nicer in the long run
* for now, `Declare Scope` should probably allow re-declaring a scope (with no warning)
* scopes in primitive numerals should behave the same as `Notation` and `Delimit Scope` (so lazily declare the scope with a warning, hopefully eventually becoming an error)
* local declaration of a scope is not so important (`Delimit Scope` and `Bind Scope` don't support `Local` currently, and I haven't seen a pressing need for it).

---

> Indeed. Conclusion is that we should deactivate the warning in Require, right?

Yes, I think that is the way to go.

> Indeed. I made a test in the standard library and it happens a couple of time that files refer to the same scope without one being directly dependent on the other (e.g. Programs.Utils and Programs.Basics, or Classes.Morphisms and Classes.CMorphisms). My opinion on these given examples is that it is legitimate to have two Declare Scope which are eventually joined. But maybe they are examples where it would not be good.

Yes, that sort of "mixin" behavior is interesting. As a general principle, I believe that requiring a single point of declaration is a good idea (if you change one name, and forget to change the other, you now accidentally created a new scope, which is one of the things we are trying to prevent in the first place).
However, that requires significant restructuring of some code, so is a more ambitious change. More on this below.

> On the other side, still in the standard library, there are independent files which defined a scope of same name without this scope intended to be the same. This is the case of pair_scope which is defined in all of ssrfun.v, FMapAVL, MSetAVL, ZNatPairs.v. Altogether, these files define 2 different notations for fst and snd and notations for two distinct notions of pairs (in prod and in a specific AVL structure). But how to ensure that two distinct libraries will never define the same scope? Should scopes be qualifiable by the library name to protect from unexpected collisions?

If I were defining the system from scratch, I would definitely want scopes to be quantifiable by the module they were declared in. At that point, you probably also want to be able to define aliases for scopes. Then, for example, you could add a module `Program.NotationScope`, require `Program.NotationScope` from `Program.Basics` and `Program.Utils`, and then in `Program.Program` you define a scope alias, so that `Program.program_scope` refers to `Program.NotationScope.program_scope`.

(You could also experiment with naming all scopes `scope`, and referring to `Program.scope` as well, especially if you can use `Local` to make `Import` not shorten the name)

I think I would also unify delimiting keys with scopes, and have `nat` be an alias for `nat_scope`, or even drop `nat_scope` entirely.

So what to do concretely. Allowing `Declare Scope` to be repeated seems the easiest choice here, and is still significant progress. Forget the whole module-scoped scopes for now, that's an unneeded large redesign.

> Probably, we shall need a Local Declare Scope (and asking that the successive Notation and co have a Local flag too).

* Currently, notations do not survive end of sections, if `Local` do not survive end of module, and if not `Local` declare scopes on `Require`, and are usable on `Import`.
* `Delimit Scope` and `Bind Scope` do not support `Local`, do not survive end of sections, declare scopes on `Require`, and take effect on `Import`.
* Then `Declare Scope` should probably not survive end of sections, declare scopes on `Require`, and do nothing on `Import`.
* If `Local` is to be supported, `Delimit Scope` and `Bind Scope` should probably also be updated, and make the command die at the end of a module (so no declaring scopes on `Require`). (and forcing following notations to be `Local`)

If we were going to have module-scoped scopes, I'd advocate for having `Require` of `Declare Scope` create a scope with the qualified name, and `Import` make it availiable with the short name unless `Local`. Then we don't have to force `Notation` in a `Local` scope to also be local, they are just bound to a qualified scope.

> Another issue, which shall have a priori to be treated in #496, is that ML printers for numerals declare notations in a scope. We have to decide whether the scope is expected to be defined in advance prior to the corresponding Declare ML Module or if the Declare ML Module does the Declare Scope itself (with some subtleties about undoing the Declare ML Module needing to undo also the Declare Scope).

I think that going forward, the scope should be declared in advance of the printer (since the printer needs to cooperate with other `Notation` and `Bind Scope`). As a transition strategy, I'd make it a warning just like `Notation`, which you then want to suppress on `Require`.
(Easier said than done, perhaps 😅)

---

@maximedenes 
> Notation clashes hit a similar limitation of the warning infrastructure w.r.t. Require. So we may want to improve that.

Aren't notation clashes only reported at `Import`, not `Require`?

Do we want to have warnings on `Require`? It seems we need to look for clashes between the current environment and what is made available when we `Require`, of which my impression was "not much and we should make it less". In that case, it might make sense to have `Require` suppress all warnings, emitting them only when the Require'd file is compiled.

**Comment by ejgallego** [comment]
> I thus submitted PR #7134 in this direction.

The PR number seems wrong @herbelin
